core/src/main/java/me/xneox/epicguard/core/storage/Database.java

```diff
@@ -52,45 +53,52 @@ public void connect() throws ClassNotFoundException {
   // Creating default table and reading addresses from the database.
   public void load() throws SQLException {
     try (var connection = this.source.getConnection(); var statement = connection.prepareStatement(
-        "CREATE TABLE IF NOT EXISTS epicguard_addresses("
-        + "`address` VARCHAR(255) NOT NULL PRIMARY KEY, "
-        + "`blacklisted` BOOLEAN NOT NULL, "
-        + "`whitelisted` BOOLEAN NOT NULL, "
-        + "`nicknames` TEXT NOT NULL"
-        + ")")) {
+            "CREATE TABLE IF NOT EXISTS epicguard_addresses("
+                    + "`address` VARCHAR(255) NOT NULL PRIMARY KEY, "
+                    + "`blacklisted` BOOLEAN NOT NULL, "
+                    + "`whitelisted` BOOLEAN NOT NULL, "
+                    + "`nicknames` TEXT NOT NULL"
+                    + ")")) {
       statement.executeUpdate();
     }
 
     try (var connection = this.source.getConnection(); var statement = connection.prepareStatement("SELECT * FROM epicguard_addresses");
-        var rs = statement.executeQuery()) {
+         var rs = statement.executeQuery()) {
 
       while (rs.next()) {
         var meta = new AddressMeta(
-            rs.getBoolean("blacklisted"),
-            rs.getBoolean("whitelisted"),
-            new ArrayList<>(Arrays.asList(rs.getString("nicknames").split(","))));
+                rs.getBoolean("blacklisted"),
+                rs.getBoolean("whitelisted"),
+                new ArrayList<>(Arrays.asList(rs.getString("nicknames").split(","))),
+                false);
 
         this.core.storageManager().addresses().put(rs.getString("address"), meta);
       }
     }
   }
 
   // Saving cached addresses to the database.
-  public void save() throws SQLException {
-    for (Map.Entry<String, AddressMeta> entry : this.core.storageManager().addresses().asMap().entrySet()) {
-      var meta = entry.getValue();
+  public synchronized void save() throws SQLException {
+    try (var connection = this.source.getConnection()) {
+      for (Map.Entry<String, AddressMeta> entry : this.core.storageManager().addresses().asMap().entrySet()) {
+        var meta = entry.getValue();
 
-      try (var connection = this.source.getConnection(); var statement = connection.prepareStatement(
-          "REPLACE INTO"
-          + " epicguard_addresses(address, blacklisted, whitelisted, nicknames)"
-          + " VALUES(?, ?, ?, ?)")) {
+        if (!meta.needsSave()) {
+          continue;
+        }
+
+        var statement = connection.prepareStatement("REPLACE INTO"
```

Comment:
Here you should use a try with resources as originally done since the object implements the AutoCloseable interface